sklearn/neighbors/tests/test_neighbors.py

```diff
@@ -163,6 +163,72 @@ def test_precomputed(random_state=42):
         assert_array_almost_equal(pred_X, pred_D)
 
 
+def test_precomputed_dense():
+    def make_train_test(X_train, X_test):
+        return (metrics.pairwise_distances(X_train),
+                metrics.pairwise_distances(X_test, X_train))
+
+    check_precomputed(make_train_test)
+
+
+@pytest.mark.parametrize('fmt', ['csr', 'lil'])
+def test_precomputed_sparse_implicit_diagonal(fmt):
+    def make_train_test(X_train, X_test):
+        nn = neighbors.NearestNeighbors(n_neighbors=3).fit(X_train)
+        return (nn.kneighbors_graph(None, mode='distance').asformat(fmt),
+                nn.kneighbors_graph(X_test, mode='distance').asformat(fmt))
+
+    check_precomputed(make_train_test, implicit_diagonal=True)
+
+
+def test_precomputed_sparse_explicit_diagonal():
+    def make_train_test(X_train, X_test):
+        nn = neighbors.NearestNeighbors(n_neighbors=3).fit(X_train)
+        return (nn.kneighbors_graph(X_train.copy(), mode='distance'),
+                nn.kneighbors_graph(X_test, mode='distance'))
+
+    check_precomputed(make_train_test, implicit_diagonal=False)
+
+
+def test_is_sorted_by_data():
+    # Test the helper function _is_sorted_by_data
+    X = csr_matrix(np.arange(10))
+    assert _is_sorted_by_data(X)
+    X[0, 2] = 5
+    assert not _is_sorted_by_data(X)
+
+
+def test_check_precomputed():
+    # Test that _check_precomputed returns a graph sorted by data
+    X = csr_matrix(np.random.RandomState(42).randn(10, 10))
+    assert not _is_sorted_by_data(X)
+    Xt = _check_precomputed(X)
+    assert _is_sorted_by_data(Xt)
+
+
+def test_precomputed_sparse_invalid():
+    # Ensures enough number of nearest neighbors
+    dist = np.array([[0., 2., 1.], [2., 0., 3.], [1., 3., 0.]])
+    dist_csr = csr_matrix(dist)
+    neigh = neighbors.NearestNeighbors(n_neighbors=1, metric="precomputed")
+    neigh.fit(dist_csr)
+    neigh.kneighbors(None, n_neighbors=2)
+    neigh.kneighbors(np.array([[0., 0., 0.]]), n_neighbors=2)
+
+    dist = np.array([[0., 2., 0.], [2., 0., 3.], [0., 3., 0.]])
+    dist_csr = csr_matrix(dist)
+    neigh.fit(dist_csr)
+    msg = "2 neighbors per samples are required, but some samples have only 1"
+    assert_raises_regex(ValueError, msg, neigh.kneighbors, None, n_neighbors=2)
+
```

Comment:
Please also check that an empty matrix throws the same error (rather than, say, getting stuck in _check_is_sorted)